### PR TITLE
Added a simplification algorithm for stair plots

### PR DESCRIPTION
**Pull request by miscco**
This is the simplification function proposed in #863.

I have also modified ACID(28) to add a use case for this simplification algorithm.

### CODE REVIEW

**Comment by miscco** [comment]
So after the discussion in #863 i added a check for monotonicity

File: src/cleanfigure.m [hunk 1]
```diff
@@ -449,6 +451,47 @@ function simplifyLine(meta, handle, targetResolution)
...
+function simplifyStairs(meta, handle)
+   % Some data might not lead to a new step in the stair, e.g.
+   % [(x_1, y_1), (x_2, y_1), ... (x_k, y_1), (x_{k+1} y_2)] 
+   % equals [(x_1, y_1), (x_{k+1} y_2)]
+
+    % Extract the data
+    xData = get(handle, 'XData');
+    yData = get(handle, 'YData');
+
+    % Dont do anything if the data is empty
```
Comment (egeerardyn) [inline src/cleanfigure.m]: Nitpick: Dont -> Don't

File: src/cleanfigure.m [hunk 2]
```diff
@@ -449,6 +451,47 @@ function simplifyLine(meta, handle, targetResolution)
...
+    if isempty(xData) || isempty(yData)
+        return;
+    end
+
+    % Check for identic data points
```
Comment (egeerardyn) [inline src/cleanfigure.m]: Nitpick: identic -> identical (or coinciding, repeated, ...)

File: src/cleanfigure.m [hunk 3]
```diff
@@ -449,6 +451,47 @@ function simplifyLine(meta, handle, targetResolution)
...
+
+    % Do not remove the last data point
+    xNoDiff(end) = false;
+    yNoDiff(end) = false;
+
+    % Identic data might lead to a visible result if the sign in the other
+    % dimension changes , e.g. the sequence  [0, 1], [0, -1], [0, 2]
+    % Therefore check for a sign change in the difference between data
+    % points (The sign changes if diff(sign)~=0)
+    xIsMonotone  = [true, diff(sign(diff(xData)))==0, true];
```
Comment (egeerardyn) [inline src/cleanfigure.m]: I think we better extract this to a function.
Moreover, I have my doubts about the correctness (either of the code or my understanding of this)

As I remember, we need to check that the whole sequence is monotonous: `x(i) <= x(j) if i < j for each i != j` (or `x(i) >= x(j) ... `)
And this function doesn't seem to do that, it rather checks for locally strict monotonicity, i.e. `x(i) < x(i+1) for each i` (or `x(i) > x(i) ...`).

Perhaps we should add some documentation in the code, e.g. some ASCII plots or data sequences which highlights what kind of points can be removed (somewhere before we actually compute anything).
```matlab
% E.g. in the plot below, the data points are indicated by `o` and `x`. We can remove the `x` points without altering the lines. So this means, either:
%  - one of the coordinates is not changing
%  - ...
%
%       x--o--+
%       |     |
% o--x--o     o--o
%
```

Comment (miscco) [inline src/cleanfigure.m]: I dont think monotony is required over the full plot. The only thing that matters is when the monotony between to consequtive pairs of points changes. thats when additional lines might occur. 

Consider your expanded example
```
%       o--x--o
%       |     |
% o--x--o     o--x--o
%       |
%       o
```
This plot is not globally monoton, but that does not mean we cannot remove data points.

Comment (egeerardyn) [inline src/cleanfigure.m]: I agree that we don't need global monotonicity to able to simplify the graph.
Sorry, I had a long day yesterday, so I think my reasoning skills were not ideal.

**Comment by miscco** [comment]
So are there any further comments on this or can i update the hashes?

**Comment by egeerardyn** [comment]
If the before/after plots look the same, you can certainly update the hashes :+1: I have no further comments (except: well done!)

**Comment by miscco** [comment]
So the diff shows that only the additional stair i introduced in the test is added.

Whoever, what i cannot grasp is why for Matlab2014a and Octave it seems to add the not simplified stair. I checked my octave 4.0 version and it recognizes the stair plot as a line plot, so that lineSimplify should take care of it. Sadly i cannot really test it. Still it is an obvious improvement for all other cases.

 [diff_simplifyStairs.zip](https://github.com/matlab2tikz/matlab2tikz/files/154457/diff_simplifyStairs.zip)

**Comment by miscco** [comment]
So given that it works and there are no further comments i would merge this tomorrow.

**Pull request merged**